### PR TITLE
CIT optional plugin - base implementation

### PR DESCRIPTION
**Pull request by apahim**
base implementation for the CIT optional plugin.

Signed-off-by: Amador Pahim <apahim@redhat.com>

### CODE REVIEW

**Comment by apahim** [comment]
@clebergnu @ldoktor, this is failing due to pylint W605, in a code not related to this patch.

File: optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        return "\n".join(out)
+
+
+class Cit(object):
+    def __init__(self, parameters, constraints):
+        self.parameters = parameters
+        # Constraints are not implemented in this example
+        self.constraints = constraints
+
+    def combine(self):
```
Comment (ldoktor) [inline optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py]: The big question here is, will the algorithm be streamable? If so this interface does not really fits and you'd have to come-up with interface to `prepare` data and then to `iterate` though data, which should allow lower memory footprint. (I do hope the final plugin will be streamable)

File: docs/source/optional_plugins/index.rst [hunk 2]
```diff
@@ -18,3 +18,4 @@ optional plugins:
    yaml_loader
    golang
    glib
+   Cit
```
Comment (ldoktor) [inline docs/source/optional_plugins/index.rst]: I don't think we should mix upper and lower cases, would you please use `cit` instead?

File: optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                 dictionaries. Each dictionary represents a combination of
+                 parameters.
+        :rtype: tuple
+        """
+        combinations = []
+        headers = [item[0] for item in self.parameters]
+        for combination in itertools.product(*(params[1]
+                                               for params in self.parameters)):
+            combinations.append(dict(zip(headers, combination)))
+        return headers, combinations
```
Comment (ldoktor) [inline optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py]: have you considered adding some example selftest?

File: docs/source/optional_plugins/Cit.rst [hunk 4]
```diff
@@ -0,0 +1,121 @@
+====================
+CIT Varianter Plugin
+====================
+
+Base Combinatorial Interaction Testing plugin. It  implementas the basic
```
Comment (clebergnu) [inline docs/source/optional_plugins/Cit.rst]: s/implementas/implements/

File: docs/source/optional_plugins/index.rst (hunk 2 above)
Comment (clebergnu) [reply docs/source/optional_plugins/index.rst]: Agreed (s/Cit/cit/)

File: optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    """
+    CIT Varianter options
+    """
+
+    name = 'cit'
+    description = "CIT Varianter options for the 'run' subcommand"
+
+    def configure(self, parser):
+
+        for name in ("run", "variants"):  # intentionally ommiting "multiplex"
```
Comment (clebergnu) [inline optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py]: It's OK to even remove the comment here about the multiplexer.

File: optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py (hunk 1 above)
Comment (apahim) [reply optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py]: Yes, agreed. But this is not real code..it's a placeholder. Let's work on this stream-able topic when the actual code arrives, shall we?

File: optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
```
[the remark this answers is not in the document]
Comment (apahim) [reply optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py]: ok

File: docs/source/optional_plugins/index.rst (hunk 2 above)
Comment (apahim) [reply docs/source/optional_plugins/index.rst]: ok

File: docs/source/optional_plugins/Cit.rst (hunk 4 above)
Comment (apahim) [reply docs/source/optional_plugins/Cit.rst]: ok

File: optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py (hunk 5 above)
Comment (apahim) [reply optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py]: ok

File: optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py (hunk 3 above)
Comment (apahim) [reply optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py]: Not at this point. Also, the actual combination code is tricky to test, considering it generates different combinations in each execution. I'd leave selftests to be created with the actual code in place.

**Pull request closed without merging**